File: CTF_Writeups/scripts_go/NovruzCTF_WYCGPRB.go
// =============================================================================
// 题目名称: Ancient Spell (SSTV)
// 题目类别: Misc - 摩尔斯电码解码
// 解题思路: 音频中使用 "bi"=点 "bo"=划 的自定义编码表示摩尔斯电码，
//
//	解码后得到明文字符串
//
// 用法: echo "bi bo bibo" | go run NovruzCTF_WYCGPRB.go
//
//	cat morse_input.txt | go run NovruzCTF_WYCGPRB.go
//
// =============================================================================
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// morseMap 标准摩尔斯电码映射表
var morseMap = map[string]string{
	".-": "A", "-...": "B", "-.-.": "C", "-..": "D", ".": "E",
	"..-.": "F", "--.": "G", "....": "H", "..": "I", ".---": "J",
	"-.-": "K", ".-..": "L", "--": "M", "-.": "N", "---": "O",
	".--.": "P", "--.-": "Q", ".-.": "R", "...": "S", "-": "T",
	"..-": "U", "...-": "V", ".--": "W", "-..-": "X", "-.--": "Y",
	"--..": "Z", "-----": "0", ".----": "1", "..---": "2", "...--": "3",
	"....-": "4", ".....": "5", "-....": "6", "--...": "7", "---..": "8",
	"----.": "9",
}

// decodeWord 将 bi/bo 编码的单词转换为摩尔斯电码再解码
// bi = 点(.)，bo = 划(-)
func decodeWord(w string) string {
	parts := []string{}
	for i := 0; i < len(w); {
		if strings.HasPrefix(w[i:], "bi") {
			parts = append(parts, ".")
			i += 2
			continue
		}
		if strings.HasPrefix(w[i:], "bo") {
			parts = append(parts, "-")
			i += 2
			continue
		}
		i++
	}
	morse := strings.Join(parts, "")
	if v, ok := morseMap[morse]; ok {
		return v
	}
	return "?"
}

func main() {
	in := bufio.NewScanner(os.Stdin)
	for in.Scan() {
		line := strings.TrimSpace(in.Text())
		if line == "" {
			continue
		}
		words := strings.Fields(line)
		out := strings.Builder{}
		for _, w := range words {
			out.WriteString(decodeWord(w))
		}
		fmt.Println(out.String())
	}
}

File: scripts_python/NovruzCTF_WYCGPRB.py
#!/usr/bin/env python3
"""
题目名称: Ancient Spell (SSTV)
类别: Misc (摩尔斯电码变体解码)
解题思路:
    1. 音频中包含自定义摩尔斯电码变体
    2. "bi" 对应点 "."，"bo" 对应划 "-"
    3. 按空格分割单词，逐个解码摩尔斯字符
    4. 从标准输入读取解码后的文本
"""

import sys

# 标准摩尔斯电码映射表
MORSE_MAP = {
    ".-": "A", "-...": "B", "-.-.": "C", "-..": "D", ".": "E",
    "..-.": "F", "--.": "G", "....": "H", "..": "I", ".---": "J",
    "-.-": "K", ".-..": "L", "--": "M", "-.": "N", "---": "O",
    ".--.": "P", "--.-": "Q", ".-.": "R", "...": "S", "-": "T",
    "..-": "U", "...-": "V", ".--": "W", "-..-": "X", "-.--": "Y",
    "--..": "Z", "-----": "0", ".----": "1", "..---": "2", "...--": "3",
    "....-": "4", ".....": "5", "-....": "6", "--...": "7", "---..": "8",
    "----.": "9",
}


def decode_word(w: str) -> str:
    """将 bi/bo 编码的单词解码为摩尔斯字符"""
    parts = []
    i = 0
    while i < len(w):
        if w.startswith("bi", i):
            parts.append(".")
            i += 2
            continue
        if w.startswith("bo", i):
            parts.append("-")
            i += 2
            continue
        i += 1
    morse = "".join(parts)
    return MORSE_MAP.get(morse, "?")


def decode_line(line: str) -> str:
    """解码一行 bi/bo 编码的文本"""
    words = line.strip().split()
    return "".join(decode_word(w) for w in words)


def main():
    """从标准输入或文件读取并解码"""
    if len(sys.argv) > 1:
        # 从文件读取
        try:
            with open(sys.argv[1], "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if line:
                        print(decode_line(line))
        except OSError as e:
            print(f"[!] 无法读取文件: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        # 从标准输入读取
        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue
            print(decode_line(line))


if __name__ == "__main__":
    main()
